keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1165 @@
+suites of integration tests, one suite per index. A failure of one suite should
+not interrupt running of other suites. Additionally, I would like to be able
+to control the maximal number of retries per index.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 1
+  runPolicy:
+    backoffLimitPerIndex: true
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+In this case, we run 10 indexes representing the test suites. We allow for one
+failure per index.
+
+#### Story 2
+
+As a CI/CD platform administrator from the [Story 1](#story-1) I want to be able
+to control the failures with the pod failure policy. In particular, I want
+to be able to use pod failure policy to avoid restarts of some indexes, based
+on exit codes.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 1
+  runPolicy:
+    backoffLimitPerIndex: true
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+  podFailurePolicy:
+    rules:
+    - action: FailIndex
+      onExitCodes:
+        operator: In
+        values: [42]
+```
+
+#### Story 3
+
+As a CI/CD platform administrator from the [Story 1](#story-1) I want to be able
+to fail the entire Job if the number of failed indexes exceeds 50%. I want to
+do this in order to cut down costs of running the tests in case of compilation
+issues that would result in all tests failing.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  runPolicy:
+    backoffLimitPerIndex: true
+    maxFailedIndexes: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+
+#### The Job object too big
+
+With the new field `.status.failedIndexes` the Job object can be significantly
+larger as every failed index is recorded in the field. For example, using the
+same format as for the `.status.completedIndexes` and allowing for completions
+equal to `10^6` the size of the field, in the worst-case scenario is 3.4MB, which
+could put significant overhead on fetching the Job object data from etcd.
```

Comment:
I assume the maximal limit on the request, and thus on the object size is 1.5mb: https://etcd.io/docs/v3.6/dev-guide/limit/.

IIUC, the only field that can grow big is `status.completedIndexes`, and its size is up to 295KB. The other fields should be relatively small (e.g. the pod failure policy was estimated to be up to 22KB in its previous KEP). 

I've written a small script that checks the new field size in the worst case scenario, depending on the number of completions, these are some example values:
- 350k - 1.17MB
- 300k - 995KB
- 250k - 819KB
- 200k - 645KB
- 150k - 470KB
- 100k - 295KB

So, if we want to push boundaries 350k should be manageable, but will make future development of Job API problematic. I think a sensible compromise might be somewhere between 100k and 300k.

I would suggest going with 200k will mean up to 940KB for the two fields, as a compromise between functionality and room for future development.

Will also ask sig-scalability folks to confirm the limit and provide their expertise.

Reply:
Yes, the per-object object size limit is 1.5M AFAIK, but overall we are trying to avoid large objects as they are problematic, especially when changed frequently.

I'm missing full context here, but design with status field growing over time seems to be quite problematic to me.

What alternatives do we have?
Can we e.g. create some dedicated object to store data we need? Or maybe infer the data based on something else?

Reply:
>Yes, the per-object object size limit is 1.5M AFAIK, but overall we are trying to avoid large objects as they are problematic, especially when changed frequently. 

The object will generally be updated only by the Job controller, in a standard sync loop, per job.

In order to limit the size I would like to suggest:
- limit completions to 10^5 if maxFailedIndexes is unspecified
- limit completions to maxFailedIndexes if specified. maxFailedIndexes  <= 10^5.

I will update the KEP with this proposal, as this will only grow the Job roughly twice, still making it safe from the 1.5MB.

> Can we e.g. create some dedicated object to store data we need? 

I think technically we can, but it will be a significant complication, also to Job users who would want to get the list of failed indexes. It would also be a different approach than for keeping succeeded (completedIndexes), which can be confusing. 

Or maybe infer the data based on something else?

We need to record the failed indexes in order to prevent the retries of them. This information isn't currently stored elsewhere.

Reply:
I also wonder about performance implications for updating this field. We will update it 10^5/10^6 times potentially, right? What's the algorithm and time complexity for one update?

Reply:
> Have you considered bitmap in hex?

Yes, but I assumed the list of the failed indexes is very important for the end-user. 

> How about dynamic format to support 10^6:

i think this would also complicate the life of end-users who write controllers to parse the list of indexes. 

However, I get the point that it would be nice to support completions up to 10^6 I think the solution I currently prefer is to constrain the number of failed indexes with `maxFailedIndexes` (needs to be <= 10^5). If you set it, then we can cope with completions being 10^6.

Reply:
> Another alternative we could consider is having an implicit limit for failures. For example, if the failures are contiguous and they can be compressed, you can continue having more failures. But if the failures are not contiguous, and we reach a limit for the length of the field, the entire job fails. Could this be an acceptable experience for job users? @jensentanlo

I think having an implicit limit. For example, by constraining the field size by 300KB may result in unpredictable behavior. Assume a user sets completions to 10^6 without controlling for maxFailedIndexes. The Job may occasionally complete (if the failed indexes are consecutive), and occasionally fail to complete if the failed indexes are non-consecutive.

I've added the alternative proposals discussed here to the Alternatives section.

Reply:
In the existing API, the worst case scenario is an Indexed Job with `completions=2e9` (max int32), `parallelism=1e5`. If the last indexes are completed in an interleaved manner, this results in: `1e5*(10+1)` characters, for a total of `~1Mi`.

We can't fit another field with the same size. Then, we need to limit  `maxFailedIndexes` to `1e4`, for a total of `~100Ki`.

Reply:
At least for me, I haven't seen a real workload that goes past 10^4 indices; as a point of comparison, AWS Batch also limits to 10^4. Maybe it would be worth surveying likely users on how big their workloads would actually be for a single indexed job; if no one uses anything larger than this becomes a theoretical discussion.

At some point, I'm guessing that the size of a cluster would come into play as well.

If it does turn out that super-large workloads are useful, then I think the key point is that users need guidance on a reasonable maximum to put. For example:

Maybe worst case is that 10^5 will fail, but the actual probability is minuscule (as confirmed by a monte-carlo simulation), then that would still be fine (with a disclaimer in the documentation). So even with the idea of failing jobs that exceed the memory limit for failed indices, we'd still need some sort of reasonable maximum.

Reply:
> 1.4MB is way too close to the limit. I think these fields should use at most one order of magnitude less than the limit. Maybe we should limit maxFailedIndexes at 10^4, when completions is greater than 10^4.

I like the idea of capping maxFailedIndexes at 10^4. I will update the KEP accordingly. Let's see what others think. 

I'm also thinking that leaving completions unlimited for Indexed Jobs is sort of omission that results in 1MB. I would prefer to limit completions in the new mode up to 10^6. Then (with also limiting the maxFailedIndexes at 10^4) it will limit `completedIndexes` at 700k and `failedIndexes` at 70k, so 770k total.

Reply:
It wasn't an omission, as we calculated the maximum size in the KEP. You never know how big people want to process indexes, and just capping parallelism was enough. But sure, capping at 10^6 for the new mode sounds reasonable. We can always relax validation later.

Reply:
I see, without capping completions we still should be fine at 1.1MB. On the other hand not capping seems like weird uncontrolled territory, when the Job size might be approaching etcd limit - complicating future additions of new fields, and putting more load on etcd. So, I'm hesitant. Let's see what others say.

For now I propose to limit the completions at 10^6 and maxFailedIndexes at 10^4, parallelism is limited at 10^5 as before.

Reply:
Update: the current proposal after some offline brainstorming of the scalability issues is that we soft-limit completions to 10^5  (with a warning above that value + user-facing documentation warning). parallelism is limited to 10^5 and maxFailedIndexes at 10^4. After the fixed estimations of the worst case scenario the Job status may take 0.44Mi at completions=10^5, and 1.56 (too much) at completions=10^6.

We still want to allow for larger completions, as it might be useful for users aware of the risk, for example:
1. when they user large completions in combination with low parallelism,
2. use reconfigured etcd to support larger objects than 1.5Mi
3. have a reason not to expect the worst case scenario for the distribution of hanging or failed indexes

Reply:
Another update after the second round of offline brainstorming: allowed parameters (completions<=10^5, parallelism<=10^5,maxFailedIndexes<=10^5) OR (completions unlimited, parallelism<=10^4, maxfailedIndexes<=10^4). More details in the KEP.  

Resolving as it seems we reached a consensensus and there haven't been any asks here for a while. Feel free to open a new comment.